docs/source-pytorch/cli/lightning_cli_advanced_3.rst

```diff
@@ -251,58 +251,88 @@ where the arguments can be passed directly through the command line without spec
 
-    from pytorch_lightning.cli import instantiate_class
+    from typing import Iterable
+    from torch.optim import Optimizer
+
+
+    OptimizerCallable = Callable[[Iterable], Optimizer]
 
 
     class MyModel(LightningModule):
-        def __init__(self, optimizer1_init: dict, optimizer2_init: dict):
+        def __init__(self, optimizer1: OptimizerCallable, optimizer2: OptimizerCallable):
             super().__init__()
-            self.optimizer1_init = optimizer1_init
-            self.optimizer2_init = optimizer2_init
+            self.optimizer1 = optimizer1
+            self.optimizer2 = optimizer2
 
         def configure_optimizers(self):
-            optimizer1 = instantiate_class(self.parameters(), self.optimizer1_init)
-            optimizer2 = instantiate_class(self.parameters(), self.optimizer2_init)
+            optimizer1 = self.optimizer1(self.parameters())
+            optimizer2 = self.optimizer2(self.parameters())
             return [optimizer1, optimizer2]
 
 
-    class MyLightningCLI(LightningCLI):
-        def add_arguments_to_parser(self, parser):
-            parser.add_optimizer_args(nested_key="optimizer1", link_to="model.optimizer1_init")
-            parser.add_optimizer_args(nested_key="optimizer2", link_to="model.optimizer2_init")
+    cli = MyLightningCLI(MyModel, auto_configure_optimizers=False)
 
+Note the type ``Callable[[Iterable], Optimizer]``, which denotes a function that receives a singe argument, some
+learnable parameters, and returns an optimizer instance. With this, from the command line it is possible to select the
+class and init arguments for each of the optimizers, as follows:
 
-    cli = MyLightningCLI(MyModel)
+.. code-block:: bash
 
-The value given to ``optimizer*_init`` will always be a dictionary including ``class_path`` and ``init_args`` entries.
-The function :func:`~pytorch_lightning.cli.instantiate_class` takes care of importing the class defined in
-``class_path`` and instantiating it using some positional arguments, in this case ``self.parameters()``, and the
-``init_args``. Any number of optimizers and learning rate schedulers can be added when using ``link_to``.
+    $ python trainer.py fit \
+        --model.optimizer1=Adam \
+        --model.optimizer1.lr=0.01 \
+        --model.optimizer2=AdamW \
+        --model.optimizer2.lr=0.0001
 
-With shorthand notation:
+In the example above, the ``OptimizerCallable`` type alias was created to illustrate what the type hint means. For
+convenience, this type alias and one for learning schedulers is available in the ``cli`` module. An example of a model
+that uses dependency injection for an optimizer and a learning scheduler is:
 
-.. code-block:: bash
+.. code-block:: python
 
-    $ python trainer.py fit \
-        --optimizer1=Adam \
-        --optimizer1.lr=0.01 \
-        --optimizer2=AdamW \
-        --optimizer2.lr=0.0001
+    from pytorch_lightning.cli import OptimizerCallable, LRSchedulerCallable, LightningCLI
```

Comment:
A good thing about LightningCLI is that the CLI logic is decoupled from the rest. The idea is that when someone implements a LightningModule, don't require any change that wouldn't make sense without the CLI. Using the dependency injection pattern for optimizers and schedulers makes sense independent from the CLI. So maybe the `OptimizerCallable` and `LRSchedulerCallable` should not be in the `cli` module.